src/mon/MgrMonitor.cc

```diff
@@ -490,7 +490,28 @@ void MgrMonitor::tick()
     return;
 
   const auto now = ceph::coarse_mono_clock::now();
-  const auto cutoff = now - std::chrono::seconds(g_conf->get_val<int64_t>("mon_mgr_beacon_grace"));
+
+  const auto mgr_beacon_grace = std::chrono::seconds(
+      g_conf->get_val<int64_t>("mon_mgr_beacon_grace"));
+
+  // Note that this is the mgr daemon's tick period, not ours (the
+  // beacon is sent with this period).
+  const auto mgr_tick_period = std::chrono::seconds(
+      g_conf->get_val<int64_t>("mgr_tick_period"));
+
+  if (last_tick != ceph::coarse_mono_clock::time_point::min()
+      && (now - last_tick > (mgr_beacon_grace - mgr_tick_period))) {
+    // This case handles either local slowness (calls being delayed
+    // for whatever reason) or cluster election slowness (a long gap
+    // between calls while an election happened)
+    dout(4) << __func__ << ": resetting beacon timeouts due to mon delay "
+            "(slow election?) of " << now - last_tick << " seconds" << dendl;
+    for (auto &i : last_beacon) {
+      i.second = now;
```

Comment:
i don't think we should be so generous to update all mgr's beacon timestamps. instead, i think we might want to use `last_tick + mgr_tick_period - mgr_beacon_grace` as the cutoff.

Reply:
Hmm, the idea is that if we have gone slow, then there are probably MgrBeacon messages queued up on peons waiting to be sent through -- the idea of resetting the last_beacons to now is that we will then be giving them mgr_beacon_grace to forward those beacons (assuming they should come soon, now that the mons are back online).

If we used an offset of `(now - last_tick) - mgr_tick_period` to account for the slowness with a specific number of seconds, then we would only be accounting for that this time through `tick()`, and we would then be strict again next time through `tick` -- the effect would be to only allow mgr_tick_period to receive the next beacons, instead of mgr_beacon_grace.

Because we see this behaviour on systems that are slow/laggy, I think it's better to use the more generous approach to allow for peons that are also being slow to forward messages after the cluster recovers (they're swapping a lot or something).

Reply:
ahh, right! thanks for your explanation. makes sense to me. the next time when the laggy mon ticks the innocent mgr(s) won't survive when the fix proposed by me. but they will with your fix.